### PR TITLE
TracerFactory implementation

### PR DESCRIPTION
**Pull request by beniamin**
- implement TracerFactory
- begin migration to the recommended library layout

### CODE REVIEW

File: src/Context/SpanContext.php [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace OpenTelemetry\Context;
+
+class SpanContext extends \OpenTelemetry\Tracing\SpanContext
```
Comment (bobstrecansky) [inline src/Context/SpanContext.php]: I think you were renaming Tracing to Trace.  Did you mean to change it here too?

File: src/Trace/Span.php [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace OpenTelemetry\Trace;
+
+class Span extends \OpenTelemetry\Tracing\Span
```
Comment (bobstrecansky) [inline src/Trace/Span.php]: I think you were renaming Tracing to Trace.  Did you mean to change it here too?

File: src/Trace/Tracer.php [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace OpenTelemetry\Trace;
+
+class Tracer extends \OpenTelemetry\Tracing\Tracer
```
Comment (bobstrecansky) [inline src/Trace/Tracer.php]: I think you were renaming Tracing to Trace.  Did you mean to change it here too?

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,13 @@
+DC_RUN_PHP = docker-compose run php
```
Comment (bobstrecansky) [inline Makefile]: This Makefile will probably make things easier.  We will probably want to update the travis.yaml to use some of these functionalities.

File: src/Trace/TracerFactory.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    protected $tracers;
+
+    /**
+     * @var SpanProcessorInterface[]
+     */
+    protected $spanProcessors;
+
+    private final function __construct()
+    {
```
Comment (bobstrecansky) [inline src/Trace/TracerFactory.php]: What are you planning on adding in this constructor?  I would assume you are planning on creating an object containing some of the included vars?

File: tests/unit/Trace/TracerFactoryTest.php [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+{
+    /**
+     * @test
+     * @expectedException Error
+     */
+    public function shouldNotBeAbleToInstantiateDirectly()
+    {
+        new TracerFactory();
+    }
+}
```
Comment (bobstrecansky) [inline tests/unit/Trace/TracerFactoryTest.php]: What other tests are you considering including for the TracerFactory?  A `getTracer` and `getInstance` test might be prudent?

File: src/Context/SpanContext.php (hunk 1 above)
Comment (beniamin) [reply src/Context/SpanContext.php]: I followed instructions from here: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/library-layout.md and I want to rename Tracing to Trace.
For now, I just extended the previous implementation to make sure that it's ok for you to rename. After we decide this is the right path, I will move the implementation here.

For this case, I moved SpanContext in OpenTelemtry\Context.

File: src/Trace/Span.php (hunk 2 above)
Comment (beniamin) [reply src/Trace/Span.php]: I followed instructions from here: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/library-layout.md and I want to rename Tracing to Trace.
For now, I just extended the previous implementation to make sure that it's ok for you to rename. After we decide this is the right path, I will move the implementation here.

File: src/Trace/Tracer.php (hunk 3 above)
Comment (beniamin) [reply src/Trace/Tracer.php]: I followed instructions from here: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/library-layout.md and I want to rename Tracing to Trace.
For now, I just extended the previous implementation to make sure that it's ok for you to rename. After we decide this is the right path, I will move the implementation here.

File: src/Trace/TracerFactory.php (hunk 5 above)
Comment (beniamin) [reply src/Trace/TracerFactory.php]: I didn't plan to do anything, just wanted to implement TracerFactory as a singleton. I took the suggestion from specification: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#obtaining-a-tracer

Now I think that I can use the constructor to inject SpanProcessors and other configuration in the TracerFactory object and not in getInstance() method.

File: tests/unit/Trace/TracerFactoryTest.php (hunk 6 above)
Comment (beniamin) [reply tests/unit/Trace/TracerFactoryTest.php]: Yes, of course. This is work in progress.

File: src/Context/SpanContext.php (hunk 1 above)
Comment (bobstrecansky) [reply src/Context/SpanContext.php]: Yeah; following the spec here is perfect.  I don't believe this existed when we implemented originally (or was inadvertent).  This change sounds good.

File: src/Trace/TracerFactory.php (hunk 5 above)
Comment (bobstrecansky) [reply src/Trace/TracerFactory.php]: Sounds like a good plan.

**Comment by bobstrecansky** [comment]
Thank you for the travisci update as well!

File: src/Trace/TracerFactory.php [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        if (self::$instance instanceof TracerFactory) {
+            return self::$instance;
+        }
+
+        $instance = new TracerFactory($spanProcessors);
+
+        return self::$instance = $instance;
+    }
+
+    public function getTracer(string $name, string $version = ""): Tracer {
```
Comment (bobstrecansky) [inline src/Trace/TracerFactory.php]: Should this `{` be on the next line?

File: examples/AlwaysOffTraceExample.php [hunk 8]
```diff
@@ -1,14 +1,15 @@
...
+use OpenTelemetry\Trace\Sampler\AlwaysOffSampler;
+use OpenTelemetry\Trace\Tracer;
+use OpenTelemetry\Trace\TracerFactory;
 
 $sampler = AlwaysOffSampler::shouldSample();
 if ($sampler) {
-    $spanContext = SpanContext::generate(); // or extract from headers
-    $tracer = new Tracer($spanContext);
+    $tracer = TracerFactory::getInstance()
+        ->getTracer('io.opentelemetry.contrib.php');
```
Comment (bobstrecansky) [inline examples/AlwaysOffTraceExample.php]: Did this naming scheme come from somewhere in particular?  I didn't see it in the spec, but I could just be missing it.

Comment (beniamin) [reply examples/AlwaysOffTraceExample.php]: Yes, I found something similar in the specs: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#obtaining-a-tracer

File: src/Trace/TracerFactory.php (hunk 7 above)
Comment (beniamin) [reply src/Trace/TracerFactory.php]: Yes, it should. I think we should add somethinh like CodeSniffer.

File: examples/AlwaysOffTraceExample.php (hunk 8 above)
Comment (bobstrecansky) [reply examples/AlwaysOffTraceExample.php]: Good enough for me!

**Pull request merged**